### PR TITLE
CP-39894: move xenopsd's daemon modules from device to service

### PR DESCRIPTION
**Pull request by psafont**
This mirrors the changes in https://github.com/xapi-project/xen-api/pull/4720 except the code that is used exclusively by the swtpm daemon, mainly the code introduced by this commit and successive modifications: [`500bbab` (#4720)](https://github.com/xapi-project/xen-api/pull/4720/commits/500bbab1e5e0bac85cf9ac73972cabb1ef8e2bf1), which I cherry picked, the modified to remove the Swtpm module and functions it depends on (because they would have been unused)

It's pretty to notice the differences between this and the swtpm branch are the removal of functions needed by the swtpm daemon; and the removal of File  in the variant `Pid.path`, which would be unused, and consequent change in behaviour in the uses of this variant.

I used `git diff feature/vtpm..private/paus/pidpaths -- ocaml/xenopsd/xc/service.ml`

For device.ml there are more differences, all the ones I found were cause because of the addition of the vtpm parameter and swtpm management as well.

Ring3 BST + BVT SR: 170941 ✅

### CODE REVIEW

File: ocaml/idl/datamodel_lifecycle.ml [hunk 1]
```diff
@@ -6,13 +6,13 @@ let prototyped_of_field = function
   | "Repository", "gpgkey_path" ->
       Some "22.12.0"
   | "host", "last_software_update" ->
-      Some "22.19.0-next"
```
Comment (lindig) [inline ocaml/idl/datamodel_lifecycle.ml:9]: Is this related or accidental?

File: ocaml/xenopsd/lib/xenops_utils.ml [hunk 2]
```diff
@@ -44,6 +44,23 @@ module Unix = struct
...
 
+(** Represent an IPC endpoint *)
+module Socket = struct
+  type t = Unix of string | Port of int
+
+  module Unix = struct
+    let path x = "unix:" ^ x
+
+    let rm x =
+      let dbg = debug "error cleaning unix socket %s: %s" x in
```
Comment (lindig) [inline ocaml/xenopsd/lib/xenops_utils.ml:55]: I think not having an explicit parameter is slightly confusing.

File: ocaml/xenopsd/xc/service.mli [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    -> restore:bool
+    -> Xenops_task.Xenops_task.task_handle
+    -> int
+    -> unit
+
+  val cancel_key : Xenctrl.domid -> Cancel_utils.key
+
+  val state_path : Xenctrl.domid -> string
+
+  val pid : xs:Xenstore.Xs.xsh -> Xenctrl.domid -> int option
```
Comment (lindig) [inline ocaml/xenopsd/xc/service.mli:66]: Does a `None` result imply that `is_running` is false? This would suggest that `is_running` is a simple test for `pid <> None`.

**Review by lindig — APPROVED** [review]

File: ocaml/idl/datamodel_lifecycle.ml (hunk 1 above)
Comment (psafont) [reply ocaml/idl/datamodel_lifecycle.ml:9]: This is a change that is forced by the versioning mechanism we have, everytime the code is compile this source file is changes, which might make it to unrelated commit if we're not careful, hence why I added it a separate commit. This is the same change as in Rob's PR

File: ocaml/xenopsd/xc/service.mli (hunk 3 above)
Comment (psafont) [reply ocaml/xenopsd/xc/service.mli:66]: Not exactly, is_running is:
```
  let is_running ~xs domid =
    match pid ~xs domid with
    | None ->
        false
    | Some p -> (
      try Unix.kill p 0 ; (* This checks the existence of pid p *)
                          true
      with _ -> false
    )
```

File: ocaml/idl/datamodel_lifecycle.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/idl/datamodel_lifecycle.ml:9]: Yeah, it was actually in 22.20.0 but got missed. I've got the right version in my PR, which was merged, so this change should now be removed.

**Review by edwintorok — APPROVED** [review]
